### PR TITLE
Allow pods to declare that they want one junit per container, instead of pod

### PR DESCRIPTION
**Pull request by smarterclayton**
Setting the annotation `ci-operator.openshift.io/container-sub-tests` to a
list of comma delimited container names will replace the overall junit
result reported for that pod or template with one junit result per test.

This allows a pod that has both setup and test containers to get per task
junits which aids debugging.

Also simplify the names of the steps as printed in junit to make them
easier to scan.

```
<testsuites>
  <testsuite name="operator" tests="1" skipped="0" failures="1" time="5.068806427">
    <testcase name="Run template template - test container test" time="0">
      <failure message="">nameserver 10.142.0.22&#xA;search clayton-test-1.svc.cluster.local svc.cluster.local cluster.local c.openshift-ci-infra.internal google.internal&#xA;options ndots:5&#xA;</failure>
    </testcase>
  </testsuite>
</testsuites>
```

for the example in the dir (time is zero because kube accuracy is +/- 1 second)

### CODE REVIEW

**Comment by smarterclayton** [comment]
This will give us accurate test grid timings of tests only, teardown only, and install only (so we can look for regressions in those components).  Will also make initial triage of failures easier.

**Comment by smarterclayton** [comment]
@bbguimaraes @petr-muller fyi

**Comment by smarterclayton** [comment]
Also @stevekuznetsov

File: pkg/steps/artifacts.go [hunk 1]
```diff
@@ -24,6 +24,7 @@ import (
 	"k8s.io/client-go/tools/remotecommand"
 
 	templateapi "github.com/openshift/api/template/v1"
+	"github.com/openshift/ci-operator/pkg/junit"
```
Comment (stevekuznetsov) [inline pkg/steps/artifacts.go]: nit: can we get this import stanza broken up correctly

File: pkg/steps/artifacts.go [hunk 2]
```diff
@@ -51,6 +52,72 @@ func (nopNotifier) Complete(_ string)               {}
 func (nopNotifier) Done(_ string) bool              { return true }
 func (nopNotifier) Cancel()                         {}
 
+// TestCaseNotifier allows a caller to generate per container unit tests
```
Comment (stevekuznetsov) [inline pkg/steps/artifacts.go]: `s/unit tests/jUnit records/`?

File: pkg/steps/artifacts.go [hunk 3]
```diff
@@ -51,6 +52,72 @@ func (nopNotifier) Complete(_ string)               {}
...
+// that provide better granularity for debugging problems. It intercepts
+// notifications and remembers the last pod retrieved.
+type TestCaseNotifier struct {
+	nested  ContainerNotifier
+	lastPod *coreapi.Pod
+}
+
+func NewTestCaseNotifier(nested ContainerNotifier) *TestCaseNotifier {
+	return &TestCaseNotifier{nested: nested}
```
Comment (stevekuznetsov) [inline pkg/steps/artifacts.go]: Why not just export the field?

File: pkg/steps/artifacts.go [hunk 4]
```diff
@@ -51,6 +52,72 @@ func (nopNotifier) Complete(_ string)               {}
...
+// SubTests returns one junit test for each terminated container with a name
+// in the annotation 'ci-operator.openshift.io/container-sub-tests' in the pod.
+// Invoking SubTests clears the last pod, so subsequent calls will return no
+// tests unless Notify() has been called in the meantime.
+func (n *TestCaseNotifier) SubTests(prefix string) []*junit.TestCase {
+	if n.lastPod == nil {
+		return nil
+	}
+	pod := n.lastPod
+	n.lastPod = nil
```
Comment (stevekuznetsov) [inline pkg/steps/artifacts.go]: Is this threadsafe?

File: pkg/steps/artifacts.go [hunk 5]
```diff
@@ -51,6 +52,72 @@ func (nopNotifier) Complete(_ string)               {}
...
+
+	names := strings.Split(pod.Annotations["ci-operator.openshift.io/container-sub-tests"], ",")
```
Comment (stevekuznetsov) [inline pkg/steps/artifacts.go]: Let's get us a constant

File: pkg/steps/artifacts.go [hunk 6]
```diff
@@ -51,6 +52,72 @@ func (nopNotifier) Complete(_ string)               {}
...
+	if len(names) == 1 && names[0] == "" {
+		return nil
+	}
+	var tests []*junit.TestCase
+	for _, status := range pod.Status.ContainerStatuses {
```
Comment (stevekuznetsov) [inline pkg/steps/artifacts.go]: No check for `""` in the comma-delimited list?

File: pkg/steps/artifacts.go [hunk 7]
```diff
@@ -51,6 +52,72 @@ func (nopNotifier) Complete(_ string)               {}
...
+		if stringInSlice(names, status.Name) {
```
Comment (stevekuznetsov) [inline pkg/steps/artifacts.go]:
```go
if sets.NewString(names...).Has(status.Name) {
```

?

File: pkg/steps/artifacts.go [hunk 8]
```diff
@@ -219,9 +286,10 @@ type ArtifactWorker struct {
 
 	podsToDownload chan string
 
-	lock      sync.Mutex
-	remaining podContainersMap
-	required  podContainersMap
+	lock         sync.Mutex
+	remaining    podContainersMap
+	required     podContainersMap
+	hasArtifacts map[string]struct{}
```
Comment (stevekuznetsov) [inline pkg/steps/artifacts.go]:
```
sets.String()
```

File: pkg/steps/artifacts.go [hunk 9]
```diff
@@ -51,6 +52,72 @@ func (nopNotifier) Complete(_ string)               {}
...
+}
+
+func (n *TestCaseNotifier) Complete(podName string)  { n.nested.Complete(podName) }
+func (n *TestCaseNotifier) Done(podName string) bool { return n.nested.Done(podName) }
+func (n *TestCaseNotifier) Cancel()                  { n.nested.Cancel() }
+
+// SubTests returns one junit test for each terminated container with a name
+// in the annotation 'ci-operator.openshift.io/container-sub-tests' in the pod.
+// Invoking SubTests clears the last pod, so subsequent calls will return no
+// tests unless Notify() has been called in the meantime.
```
Comment (stevekuznetsov) [inline pkg/steps/artifacts.go]: Not sure I fully understand the point of this caching mechanism

File: test/template.yaml [hunk 10]
```diff
@@ -6,6 +6,8 @@ objects:
   apiVersion: v1
   metadata:
     name: test
+    annotations:
+      ci-operator.openshift.io/container-sub-tests: test
```
Comment (stevekuznetsov) [inline test/template.yaml]: Is it possible for us to write a test that would show us this behavior actually happening?

File: pkg/steps/artifacts.go (hunk 3 above)
Comment (smarterclayton) [reply pkg/steps/artifacts.go]: Just to be consistent with the others

File: pkg/steps/artifacts.go (hunk 4 above)
Comment (smarterclayton) [reply pkg/steps/artifacts.go]: Today we only use it in single threaded contexts, there's some affordances built in to tolerate multithreaded notification (basically setup single thread, single threaded notify, parallel upload but doesn't read these fields, then single threaded junit extraction).   I can add comments to which methods are expected to be thread safe.

File: pkg/steps/artifacts.go (hunk 6 above)
Comment (smarterclayton) [reply pkg/steps/artifacts.go]: I can, it couldn't cause a match (since we wouldn't find a container with "")

File: pkg/steps/artifacts.go (hunk 9 above)
Comment (smarterclayton) [reply pkg/steps/artifacts.go]: We use the same notifier for multiple pods in a for loop in template processing.  So if you have template with pods A and B, we pass the same notifier to A that wee do to B.  That's also how artifacts are gathered right now (we wait until A passes, then check B).

File: test/template.yaml (hunk 10 above)
Comment (smarterclayton) [reply test/template.yaml]: Probably.

File: pkg/steps/artifacts.go (hunk 9 above)
Comment (stevekuznetsov) [reply pkg/steps/artifacts.go]: I see

**Comment by smarterclayton** [comment]
Comments addressed

**Comment by stevekuznetsov** [comment]
/lgtm
/hold
@petr-muller want to take a look?

**Comment by petr-muller** [comment]
I'll look today

**Pull request merged**